src/bridgetree.rs

```diff
@@ -86,107 +105,140 @@ impl From<Position> for usize {
     }
 }
 
+/// A set of leaves of a Merkle tree. 
 #[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
 pub enum Leaf<A> {
     Left(A),
     Right(A, A),
 }
 
-#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
-pub struct Parent<A> {
-    value: A,
-}
-
+/// A `[NonEmptyFrontier]` is a reduced representation of a Merkle tree,
+/// having either one or two leaf values, and then a set of hashes produced
+/// by the reduction of previously appended leaf values. 
 #[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
 pub struct NonEmptyFrontier<H> {
     position: Position,
     leaf: Leaf<H>,
-    parents: Vec<Parent<H>>,
+    ommers: Vec<H>,
 }
 
-impl<H: Hashable + Clone> NonEmptyFrontier<H> {
+impl<H> NonEmptyFrontier<H> {
+    /// Constructs a new frontier with the specified value at position 0.
     pub fn new(value: H) -> Self {
         NonEmptyFrontier {
             position: Position::zero(),
             leaf: Leaf::Left(value),
-            parents: vec![],
+            ommers: vec![],
         }
     }
 
+    /// Returns the altitude of the highest ommer in the frontier.
+    pub fn max_altitude(&self) -> Altitude {
+        self.position.max_altitude()
+    }
+
+    /// Returns the position of the most recently appended leaf.
+    pub fn position(&self) -> Position {
+        self.position
+    }
+
+    /// Returns the number of leaves that have been appended to this frontier.
+    pub fn size(&self) -> usize {
+        self.position.0 + 1
+    }
+}
+
+impl<H: Clone> NonEmptyFrontier<H> {
+    /// Returns the value of the leaf or ommer at the specified altitude.
+    pub fn value_at(&self, lvl: Altitude) -> Option<H> {
+        if lvl == Altitude::zero() {
+            Some(self.leaf_value())
+        } else {
+            self.ommers
+                .iter()
+                .zip(self.position.ommer_altitudes())
+                .find(|(_, l)| *l == lvl)
+                .map(|(p, _)| p.clone())
+        }
+    }
+
+    /// Returns the value of the most recently appended leaf.
+    pub fn leaf_value(&self) -> H {
+        match &self.leaf {
+            Leaf::Left(v) => v.clone(),
+            Leaf::Right(_, v) => v.clone(),
```

Comment:
Untested, but I think this works (and conveys a bit better that we are just performing value extraction):
```suggestion
            Leaf::Left(v) | Leaf::Right(_, v) => v.clone(),
```